README.md
## Hi there :heart:, I'm wanghaisheng

<img width="30%" src="https://avatars.githubusercontent.com/u/2363295?v=4" alt="wanghaisheng"/>

## Statistics

> Automated deployment @ 2023-06-19 08:11:53 Asia/Shanghai
### User

- **Total_stars_earned**: 3082

- **Projects**: 1621

### Sources

- **count**: 320

| **name** | **stars** | **forks** | **pushed_at** |
| :---- | :---- | :---- | :---- |
|[wanghaisheng/gen-brand-social-account-RSS](https://github.com/wanghaisheng/gen-brand-social-account-RSS)|2|3|2023-06-18T06:01:18Z|
|[wanghaisheng/capcut-tweets-capture](https://github.com/wanghaisheng/capcut-tweets-capture)|0|0|2023-06-18T00:14:39Z|
|[wanghaisheng/tiktok-opensdk-web](https://github.com/wanghaisheng/tiktok-opensdk-web)|0|0|2023-06-17T19:04:07Z|
|[wanghaisheng/longtail-keywords-expand-tools-GUI](https://github.com/wanghaisheng/longtail-keywords-expand-tools-GUI)|1|0|2023-06-16T09:13:12Z|
|[wanghaisheng/auto-gen-apple-ids](https://github.com/wanghaisheng/auto-gen-apple-ids)|0|0|2023-06-15T11:24:47Z|
|[wanghaisheng/tiktoka-studio-uploader](https://github.com/wanghaisheng/tiktoka-studio-uploader)|101|28|2023-06-15T06:39:09Z|
|[wanghaisheng/validate-your-proxy-against-ip-leak](https://github.com/wanghaisheng/validate-your-proxy-against-ip-leak)|0|0|2023-06-15T06:34:54Z|
|[wanghaisheng/tiktoka-studio-uploader-landingpage-bigspring](https://github.com/wanghaisheng/tiktoka-studio-uploader-landingpage-bigspring)|0|0|2023-06-15T00:52:09Z|
|[wanghaisheng/tiktoka-studio-landingpage](https://github.com/wanghaisheng/tiktoka-studio-landingpage)|0|0|2023-06-15T00:47:56Z|
|[wanghaisheng/capcutlab](https://github.com/wanghaisheng/capcutlab)|0|0|2023-06-14T07:53:23Z|
|[wanghaisheng/ask-temu-blog](https://github.com/wanghaisheng/ask-temu-blog)|0|0|2023-06-14T07:21:39Z|
|[wanghaisheng/ask-temu-website](https://github.com/wanghaisheng/ask-temu-website)|0|0|2023-06-14T07:18:52Z|
|[wanghaisheng/video-publish-metadata-gpt](https://github.com/wanghaisheng/video-publish-metadata-gpt)|0|1|2023-06-13T05:29:57Z|
|[wanghaisheng/idea-to-action-cycle](https://github.com/wanghaisheng/idea-to-action-cycle)|0|0|2023-06-12T11:38:21Z|
|[wanghaisheng/astro-microsoft-clarity](https://github.com/wanghaisheng/astro-microsoft-clarity)|0|0|2023-06-12T04:00:00Z|
|[wanghaisheng/100-mini-project-before-40-Odyssey](https://github.com/wanghaisheng/100-mini-project-before-40-Odyssey)|0|0|2023-06-11T07:54:22Z|
|[wanghaisheng/ttsall](https://github.com/wanghaisheng/ttsall)|0|0|2023-06-10T19:02:02Z|
|[wanghaisheng/awesome-wearable-device](https://github.com/wanghaisheng/awesome-wearable-device)|1|0|2023-06-10T16:26:50Z|
|[wanghaisheng/get-tiktok-user-video-list](https://github.com/wanghaisheng/get-tiktok-user-video-list)|7|1|2023-06-10T09:01:11Z|
|[wanghaisheng/arxiv-paper-daily](https://github.com/wanghaisheng/arxiv-paper-daily)|2|2|2023-06-10T08:59:08Z|
|[wanghaisheng/heart-rate-arxiv-paper](https://github.com/wanghaisheng/heart-rate-arxiv-paper)|0|0|2023-06-10T07:11:06Z|
|[wanghaisheng/100-mini-project-before-40-astrowind](https://github.com/wanghaisheng/100-mini-project-before-40-astrowind)|0|0|2023-06-10T07:04:43Z|
|[wanghaisheng/wearable-device-arxiv-paper-daily](https://github.com/wanghaisheng/wearable-device-arxiv-paper-daily)|0|0|2023-06-09T14:10:35Z|
|[wanghaisheng/personal-interest-paper-daily](https://github.com/wanghaisheng/personal-interest-paper-daily)|0|0|2023-06-09T12:53:19Z|
|[wanghaisheng/astro-seo-all-in-one](https://github.com/wanghaisheng/astro-seo-all-in-one)|0|0|2023-06-09T10:59:47Z|
|[wanghaisheng/arxiv-paper-daily-astro](https://github.com/wanghaisheng/arxiv-paper-daily-astro)|0|0|2023-06-09T09:07:01Z|
|[wanghaisheng/ocr-arxiv-daily](https://github.com/wanghaisheng/ocr-arxiv-daily)|12|0|2023-06-07T21:52:57Z|
|[wanghaisheng/github-action-scihub-download](https://github.com/wanghaisheng/github-action-scihub-download)|0|0|2023-06-07T12:34:56Z|
|[wanghaisheng/flybirds-gpt](https://github.com/wanghaisheng/flybirds-gpt)|0|0|2023-06-06T07:03:26Z|
|[wanghaisheng/pyodide-playground](https://github.com/wanghaisheng/pyodide-playground)|0|0|2023-06-04T16:07:50Z|
|[wanghaisheng/flet-saas-starter](https://github.com/wanghaisheng/flet-saas-starter)|2|1|2023-06-02T14:18:37Z|
|[wanghaisheng/goodread-review-to-video](https://github.com/wanghaisheng/goodread-review-to-video)|0|0|2023-06-01T11:44:20Z|
|[wanghaisheng/wanghaisheng](https://github.com/wanghaisheng/wanghaisheng)|0|0|2023-05-31T02:23:27Z|
|[wanghaisheng/TiktokaStudioDownload](https://github.com/wanghaisheng/TiktokaStudioDownload)|1|0|2023-05-31T01:42:38Z|
|[wanghaisheng/-tauri-vue-template-demo](https://github.com/wanghaisheng/-tauri-vue-template-demo)|0|0|2023-05-25T12:55:22Z|
|[wanghaisheng/nextjs-blog-theme](https://github.com/wanghaisheng/nextjs-blog-theme)|0|0|2023-05-13T15:37:52Z|
|[wanghaisheng/accessible-astro-dashboard](https://github.com/wanghaisheng/accessible-astro-dashboard)|0|0|2023-05-11T02:50:10Z|
|[wanghaisheng/bulk-download-transcript-from-youtube-channel](https://github.com/wanghaisheng/bulk-download-transcript-from-youtube-channel)|0|0|2023-05-09T07:36:11Z|
|[wanghaisheng/bulk-download-youtube-channel-videos-transcription](https://github.com/wanghaisheng/bulk-download-youtube-channel-videos-transcription)|0|0|2023-05-09T03:39:35Z|
|[wanghaisheng/resume-parse-evaluation](https://github.com/wanghaisheng/resume-parse-evaluation)|141|42|2023-05-08T08:42:08Z|
|[wanghaisheng/can-gpt-build-daily-agent](https://github.com/wanghaisheng/can-gpt-build-daily-agent)|0|0|2023-05-08T02:36:05Z|
|[wanghaisheng/evaluation-of-gpt-translation-of-tech-doc-from-their-website](https://github.com/wanghaisheng/evaluation-of-gpt-translation-of-tech-doc-from-their-website)|0|0|2023-05-07T10:01:49Z|
|[wanghaisheng/ChatGPT-On-WeChat](https://github.com/wanghaisheng/ChatGPT-On-WeChat)|0|0|2023-04-21T16:51:13Z|
|[wanghaisheng/build-docker-for-serverless-deploy-starter](https://github.com/wanghaisheng/build-docker-for-serverless-deploy-starter)|1|0|2023-03-13T02:19:07Z|
|[wanghaisheng/electron-vite-vue-auto-video-publish](https://github.com/wanghaisheng/electron-vite-vue-auto-video-publish)|0|0|2023-03-01T09:01:16Z|
|[wanghaisheng/brandninja](https://github.com/wanghaisheng/brandninja)|1|0|2023-01-11T21:06:46Z|
|[wanghaisheng/awesome-microservice](https://github.com/wanghaisheng/awesome-microservice)|482|127|2022-11-30T11:15:29Z|
|[wanghaisheng/tiktok-trending-api-data-archive](https://github.com/wanghaisheng/tiktok-trending-api-data-archive)|3|0|2022-11-15T00:22:54Z|
|[wanghaisheng/citespace-auto](https://github.com/wanghaisheng/citespace-auto)|0|0|2022-07-21T12:58:03Z|
|[wanghaisheng/Door-Prints-in-china](https://github.com/wanghaisheng/Door-Prints-in-china)|1|0|2022-07-13T21:17:36Z|
|[wanghaisheng/monster-story-in-china](https://github.com/wanghaisheng/monster-story-in-china)|0|0|2022-07-13T20:56:31Z|
|[wanghaisheng/comment-of-2dog-interviews](https://github.com/wanghaisheng/comment-of-2dog-interviews)|0|0|2022-07-08T16:51:41Z|
|[wanghaisheng/tcm-beauty](https://github.com/wanghaisheng/tcm-beauty)|1|0|2022-07-08T11:27:40Z|
|[wanghaisheng/jiang-zhuang-nai-review](https://github.com/wanghaisheng/jiang-zhuang-nai-review)|0|0|2022-07-08T10:51:27Z|
|[wanghaisheng/diet-treatment-tcm](https://github.com/wanghaisheng/diet-treatment-tcm)|2|0|2022-07-08T10:11:58Z|
|[wanghaisheng/building-great-team](https://github.com/wanghaisheng/building-great-team)|1|0|2022-07-08T10:09:01Z|
|[wanghaisheng/social-media-monitor-weekly-report](https://github.com/wanghaisheng/social-media-monitor-weekly-report)|3|1|2022-07-08T03:52:48Z|
|[wanghaisheng/xiaogaojie](https://github.com/wanghaisheng/xiaogaojie)|2|0|2022-07-07T15:55:27Z|
|[wanghaisheng/is-3p-against-law](https://github.com/wanghaisheng/is-3p-against-law)|0|0|2022-07-07T11:25:32Z|
|[wanghaisheng/Alopecia-solution](https://github.com/wanghaisheng/Alopecia-solution)|1|0|2022-07-06T18:59:48Z|
|[wanghaisheng/xiaochidian-research](https://github.com/wanghaisheng/xiaochidian-research)|0|0|2022-07-06T14:42:27Z|
|[wanghaisheng/airtest-record-demo](https://github.com/wanghaisheng/airtest-record-demo)|0|0|2022-07-06T04:36:57Z|
|[wanghaisheng/xiaogaojie-recipes](https://github.com/wanghaisheng/xiaogaojie-recipes)|0|0|2022-07-05T18:11:37Z|
|[wanghaisheng/make-video-no-code](https://github.com/wanghaisheng/make-video-no-code)|1|0|2022-07-05T13:03:11Z|
|[wanghaisheng/app-review-csv-to-webgal-scripts](https://github.com/wanghaisheng/app-review-csv-to-webgal-scripts)|1|0|2022-07-05T12:36:24Z|
|[wanghaisheng/smart-home-sleep-disorder-scene](https://github.com/wanghaisheng/smart-home-sleep-disorder-scene)|0|0|2022-07-01T00:17:37Z|
|[wanghaisheng/tinnitus-dtx](https://github.com/wanghaisheng/tinnitus-dtx)|3|0|2022-06-30T23:31:31Z|
|[wanghaisheng/Cognitive-Behavioral-Therapy](https://github.com/wanghaisheng/Cognitive-Behavioral-Therapy)|1|0|2022-06-30T22:31:34Z|
|[wanghaisheng/evidence-based-health-style-for-ederly](https://github.com/wanghaisheng/evidence-based-health-style-for-ederly)|0|0|2022-06-30T21:54:35Z|
|[wanghaisheng/A-Survey-on-Wearable-Technology-History-State-of-the-Art-and-Current-Challenges](https://github.com/wanghaisheng/A-Survey-on-Wearable-Technology-History-State-of-the-Art-and-Current-Challenges)|1|0|2022-06-30T21:20:56Z|
|[wanghaisheng/ederly-design](https://github.com/wanghaisheng/ederly-design)|0|0|2022-06-30T20:49:01Z|
|[wanghaisheng/high-net-worth-client-like](https://github.com/wanghaisheng/high-net-worth-client-like)|0|0|2022-06-30T20:31:49Z|
|[wanghaisheng/truth-of-modern-business-model](https://github.com/wanghaisheng/truth-of-modern-business-model)|1|0|2022-06-30T10:40:42Z|
|[wanghaisheng/get-involved-template](https://github.com/wanghaisheng/get-involved-template)|0|0|2022-06-29T16:44:24Z|
|[wanghaisheng/jianshen-qigong](https://github.com/wanghaisheng/jianshen-qigong)|1|0|2022-06-29T04:30:41Z|
|[wanghaisheng/sleep-analysis-iot](https://github.com/wanghaisheng/sleep-analysis-iot)|0|0|2022-06-29T04:04:17Z|
|[wanghaisheng/wearable-paper-to-video](https://github.com/wanghaisheng/wearable-paper-to-video)|0|0|2022-06-29T03:04:18Z|
|[wanghaisheng/-Gift-Tiktok-Live-flybirds](https://github.com/wanghaisheng/-Gift-Tiktok-Live-flybirds)|0|0|2022-06-28T21:30:34Z|
|[wanghaisheng/copycat-account-detect](https://github.com/wanghaisheng/copycat-account-detect)|1|0|2022-06-28T10:55:40Z|
|[wanghaisheng/living-will-fhir](https://github.com/wanghaisheng/living-will-fhir)|1|0|2022-06-27T14:05:45Z|
|[wanghaisheng/start-micro-saas-0-1](https://github.com/wanghaisheng/start-micro-saas-0-1)|0|0|2022-06-24T16:02:47Z|
|[wanghaisheng/tiktoka-studio-uploader-app](https://github.com/wanghaisheng/tiktoka-studio-uploader-app)|1|0|2022-06-24T10:42:00Z|
|[wanghaisheng/awesome-ocr](https://github.com/wanghaisheng/awesome-ocr)|1637|356|2022-06-24T04:50:59Z|
|[wanghaisheng/url2video-pdf](https://github.com/wanghaisheng/url2video-pdf)|1|0|2022-06-23T19:25:31Z|
|[wanghaisheng/ocr-tweets-monitoring](https://github.com/wanghaisheng/ocr-tweets-monitoring)|0|0|2022-06-23T09:33:16Z|
|[wanghaisheng/ocr-paper-collector](https://github.com/wanghaisheng/ocr-paper-collector)|4|0|2022-06-23T07:50:45Z|
|[wanghaisheng/recipe-of-xiaogaojie](https://github.com/wanghaisheng/recipe-of-xiaogaojie)|0|0|2022-06-22T20:18:47Z|
|[wanghaisheng/fhir-cn](https://github.com/wanghaisheng/fhir-cn)|71|23|2022-06-22T16:45:03Z|
|[wanghaisheng/gastritis](https://github.com/wanghaisheng/gastritis)|0|0|2022-06-22T04:35:56Z|
|[wanghaisheng/live-streaming-transcript-dataset](https://github.com/wanghaisheng/live-streaming-transcript-dataset)|2|2|2022-06-22T04:14:20Z|
|[wanghaisheng/lawn-mowing-video-website](https://github.com/wanghaisheng/lawn-mowing-video-website)|1|0|2022-06-21T19:38:43Z|
|[wanghaisheng/make-xianjian-video](https://github.com/wanghaisheng/make-xianjian-video)|1|0|2022-06-21T19:20:30Z|
|[wanghaisheng/make-rick-and-morty-style-video](https://github.com/wanghaisheng/make-rick-and-morty-style-video)|1|0|2022-06-21T19:11:07Z|
|[wanghaisheng/creator-email-collector-gui-trial](https://github.com/wanghaisheng/creator-email-collector-gui-trial)|0|0|2022-06-20T22:38:50Z|
|[wanghaisheng/tech-stack-for-cross-platform-app](https://github.com/wanghaisheng/tech-stack-for-cross-platform-app)|0|0|2022-06-20T19:23:11Z|
|[wanghaisheng/python-desktop-app-amis-starter](https://github.com/wanghaisheng/python-desktop-app-amis-starter)|1|0|2022-06-20T11:37:20Z|
|[wanghaisheng/WebGAL_Live_Demo](https://github.com/wanghaisheng/WebGAL_Live_Demo)|2|0|2022-06-18T11:56:42Z|
|[wanghaisheng/jelly-master](https://github.com/wanghaisheng/jelly-master)|0|0|2022-06-17T02:39:29Z|
|[wanghaisheng/ai-chips-community-growth-strategy](https://github.com/wanghaisheng/ai-chips-community-growth-strategy)|1|1|2022-06-16T15:12:01Z|
|[wanghaisheng/Irritable-Bowel-Syndrome](https://github.com/wanghaisheng/Irritable-Bowel-Syndrome)|0|0|2022-06-15T19:36:37Z|
|[wanghaisheng/nocturnal-enuresis](https://github.com/wanghaisheng/nocturnal-enuresis)|0|0|2022-06-14T21:58:43Z|
|[wanghaisheng/dtx-notes](https://github.com/wanghaisheng/dtx-notes)|0|0|2022-06-14T20:48:38Z|
|[wanghaisheng/how-disposable-diapers-hurt-baby](https://github.com/wanghaisheng/how-disposable-diapers-hurt-baby)|0|0|2022-06-14T19:53:14Z|
|[wanghaisheng/tiktoka-studio-website](https://github.com/wanghaisheng/tiktoka-studio-website)|0|0|2022-06-13T17:31:56Z|
|[wanghaisheng/remake-of-mendel.ai](https://github.com/wanghaisheng/remake-of-mendel.ai)|1|0|2022-06-12T09:51:35Z|
|[wanghaisheng/ridiculous-web5](https://github.com/wanghaisheng/ridiculous-web5)|1|0|2022-06-11T17:49:57Z|
|[wanghaisheng/track-2b-customer-true-requirements](https://github.com/wanghaisheng/track-2b-customer-true-requirements)|1|0|2022-06-11T16:07:10Z|
|[wanghaisheng/DevRel-community-building-Growth-toolkit](https://github.com/wanghaisheng/DevRel-community-building-Growth-toolkit)|0|0|2022-06-11T13:22:46Z|
|[wanghaisheng/copycat-of-youtube-comments](https://github.com/wanghaisheng/copycat-of-youtube-comments)|0|0|2022-06-09T12:49:42Z|
|[wanghaisheng/pywebio-fastapi-micro-saas-starter](https://github.com/wanghaisheng/pywebio-fastapi-micro-saas-starter)|0|0|2022-05-31T07:58:23Z|
|[wanghaisheng/copyright-free-catalog](https://github.com/wanghaisheng/copyright-free-catalog)|0|0|2022-05-30T23:19:36Z|
|[wanghaisheng/scrape-sprites](https://github.com/wanghaisheng/scrape-sprites)|0|0|2022-05-29T23:12:08Z|
|[wanghaisheng/awesome-product-journey-tool](https://github.com/wanghaisheng/awesome-product-journey-tool)|0|0|2022-05-27T03:25:01Z|
|[wanghaisheng/vue-amis-editor](https://github.com/wanghaisheng/vue-amis-editor)|0|1|2022-05-27T01:20:37Z|
|[wanghaisheng/github-journey](https://github.com/wanghaisheng/github-journey)|0|0|2022-05-26T16:24:10Z|
|[wanghaisheng/lp-11](https://github.com/wanghaisheng/lp-11)|0|0|2022-05-26T06:44:18Z|
|[wanghaisheng/chrome-extension-for-shopify-spy](https://github.com/wanghaisheng/chrome-extension-for-shopify-spy)|0|0|2022-05-26T06:40:11Z|
|[wanghaisheng/youtube-automation-toolkit](https://github.com/wanghaisheng/youtube-automation-toolkit)|4|2|2022-05-26T05:08:38Z|
|[wanghaisheng/flybirds-automation-starter](https://github.com/wanghaisheng/flybirds-automation-starter)|0|0|2022-05-26T02:19:44Z|
|[wanghaisheng/ace-attorney-story-video-auto-generation](https://github.com/wanghaisheng/ace-attorney-story-video-auto-generation)|2|0|2022-05-23T16:46:12Z|
|[wanghaisheng/Sync-YouTube-Podcast](https://github.com/wanghaisheng/Sync-YouTube-Podcast)|1|0|2022-05-23T04:10:42Z|
|[wanghaisheng/YouTube-Podcast](https://github.com/wanghaisheng/YouTube-Podcast)|1|0|2022-05-23T03:48:10Z|
|[wanghaisheng/next-lead-generation](https://github.com/wanghaisheng/next-lead-generation)|0|0|2022-05-22T04:24:24Z|
|[wanghaisheng/indiehacker-2-video-audiobook-ebook-pdf](https://github.com/wanghaisheng/indiehacker-2-video-audiobook-ebook-pdf)|0|0|2022-05-19T06:54:02Z|
|[wanghaisheng/tiktoka-sponsors](https://github.com/wanghaisheng/tiktoka-sponsors)|1|0|2022-05-19T04:48:00Z|
|[wanghaisheng/.github](https://github.com/wanghaisheng/.github)|0|0|2022-05-18T11:09:56Z|
|[wanghaisheng/worker-kit-email](https://github.com/wanghaisheng/worker-kit-email)|1|0|2022-05-17T08:46:24Z|
|[wanghaisheng/twitch-bot-worker](https://github.com/wanghaisheng/twitch-bot-worker)|1|0|2022-05-16T16:10:40Z|
|[wanghaisheng/tiktoka-video-copyright-infringement-auto](https://github.com/wanghaisheng/tiktoka-video-copyright-infringement-auto)|0|0|2022-05-15T14:40:14Z|
|[wanghaisheng/moviepy-safe-delete](https://github.com/wanghaisheng/moviepy-safe-delete)|0|0|2022-05-13T10:42:55Z|
|[wanghaisheng/sponsor-Website](https://github.com/wanghaisheng/sponsor-Website)|0|0|2022-05-12T13:41:17Z|
|[wanghaisheng/link-detec-shopconan](https://github.com/wanghaisheng/link-detec-shopconan)|0|0|2022-05-08T17:22:14Z|
|[wanghaisheng/reliable](https://github.com/wanghaisheng/reliable)|1|0|2022-05-08T12:23:07Z|
|[wanghaisheng/USWDS-example](https://github.com/wanghaisheng/USWDS-example)|0|0|2022-05-06T06:01:20Z|
|[wanghaisheng/GROBID-serverless](https://github.com/wanghaisheng/GROBID-serverless)|1|0|2022-05-05T16:37:52Z|
|[wanghaisheng/ai-chip-paper-and-showcase](https://github.com/wanghaisheng/ai-chip-paper-and-showcase)|1|1|2022-05-04T22:48:37Z|
|[wanghaisheng/inuchip-showcase](https://github.com/wanghaisheng/inuchip-showcase)|0|1|2022-05-04T07:54:58Z|
|[wanghaisheng/dockerfile](https://github.com/wanghaisheng/dockerfile)|0|0|2022-04-28T14:18:10Z|
|[wanghaisheng/code_saas_idea_hunter](https://github.com/wanghaisheng/code_saas_idea_hunter)|0|0|2022-04-26T14:26:49Z|
|[wanghaisheng/hongsedianyingzimu](https://github.com/wanghaisheng/hongsedianyingzimu)|1|0|2022-04-24T11:30:23Z|
|[wanghaisheng/subscription-based-saas](https://github.com/wanghaisheng/subscription-based-saas)|1|1|2022-04-23T02:13:39Z|

changelog.md
1.https://shopconna.com/showcase/
一个repo生成一个的showcase md
首页用这个showcase来展现


main.py
"""
('avatar_url', 'https://avatars.githubusercontent.com/u/2363295?s=96&v=4')
('gravatar_id', '')
('url', 'https://api.github.com/users/wanghaisheng')
('html_url', 'https://github.com/wanghaisheng')
('followers_url', 'https://api.github.com/users/wanghaisheng/followers')
('following_url', 'https://api.github.com/users/wanghaisheng/following{/other_user}')
('gists_url', 'https://api.github.com/users/wanghaisheng/gists{/gist_id}')
('starred_url', 'https://api.github.com/users/wanghaisheng/starred{/owner}{/repo}')
('subscriptions_url', 'https://api.github.com/users/wanghaisheng/subscriptions')
('organizations_url', 'https://api.github.com/users/wanghaisheng/orgs')
('repos_url', 'https://api.github.com/users/wanghaisheng/repos')
('events_url', 'https://api.github.com/users/wanghaisheng/events{/privacy}')
('received_events_url', 'https://api.github.com/users/wanghaisheng/received_events')
('issues_url', 'https://api.github.com/search/issues?q=type:pr+is:merged+author:wanghaisheng&per_page=100')
"""
# -*- coding: utf-8 -*-
# Time       : 2022/5/11 18:48
# Author     : QIN2DIM
# Github     : https://github.com/QIN2DIM
# Description:
import os
from datetime import datetime
from typing import Union, Optional, List

import requests

GITHUB_ACTOR = os.getenv("GITHUB_ACTOR", "QIN2DIM")

PATH_SUMMARY_OUTPUT = "README.md"

API_ROOT = "https://api.github.com"
USER_ROOT = f"{API_ROOT}/users/{GITHUB_ACTOR}"
REPO_URLS = f"{USER_ROOT}/repos"
EVENT_URLS = f"{USER_ROOT}/events"
ISSUE_URLS = f"{API_ROOT}/search/issues"


class GitUserStatus:
    def __init__(self):
        self.user_obj = {}
        self.repo_objs = {}
        self.forked_repo_objs = {}
        self.contributed_repo_objs = {}

        self.repo2hyperlink = {}

        # User basic
        self.avatar_url = "https://avatars.githubusercontent.com/u/000000?v=4"
        self.public_repos = 0

        # User statistics
        self.total_stars_earned = 0
        self.total_commits = 0
        self.total_prs = 0
        self.total_issues = 0
        self.contributed_to = 0

        # Enter api root
        self._init_user_info()

        # Split tasks according to the number of projects.
        # 30 <= per_page <= 100
        self.per_page = 100
        self.pages = (
            1 if self.public_repos <= self.per_page else int(self.public_repos / self.per_page) + 1
        )

    def _init_user_info(self):
        self.user_obj.update(requests.get(USER_ROOT).json())

        self.avatar_url = self.user_obj.get("avatar_url", self.avatar_url)
        self.public_repos = self.user_obj.get("public_repos", self.public_repos)

    def get_repos_info(self, patch_statistics: bool = True):
        """获取仓库 stars forks watches"""
        # 解析用户仓库对象
        for index, page in enumerate(range(1, self.pages + 1)):
            url = f"{REPO_URLS}?page={page}&&per_page={self.per_page}"
            repos = requests.get(url).json()
            for repo in repos:
                # 仅统计原创项目
                _docker = self.forked_repo_objs if repo["fork"] else self.repo_objs
                _docker[repo["full_name"]] = {
                    "stars": repo["stargazers_count"],
                    "forks": repo["forks_count"],
                    # "watches": repo["watchers_count"], # 失效
                    "pushed_at": repo["pushed_at"],
                }
                # 获取仓库超链接映射
                self.repo2hyperlink[repo["full_name"]] = repo["html_url"]
                # 補充用戶統計信息
                if patch_statistics and not repo["fork"]:
                    self.total_stars_earned += repo["stargazers_count"]

    def get_contributed_repos(self):
        skip_repo = []
        content = []
        repo2stars = {}

        for _, page in enumerate(range(1, 10)):
            url = f"{ISSUE_URLS}?q=type:pr+is:merged+author:{GITHUB_ACTOR}&page={page}&per_page=100"
            events = requests.get(url).json()

            # Ending Unexpected Retrieval Tasks.
            if not events.get("items"):
                break
            content.extend(events["items"])

        for event in reversed(content):
            repo_owner = event["repository_url"].split("/")[-2]
            repo_name = f'{repo_owner}/{event["repository_url"].split("/")[-1]}'
            repo_api = event["repository_url"]

            # Strategy: Skip task. Filtering your own project.
            if repo_owner.lower() == GITHUB_ACTOR.lower():
                continue
            # Strategy: Skip task. API response is outdated, skip task.
            if repo_name in skip_repo:
                continue
            # Strategy: Skip task. Filtering duplicate request tasks.
            if repo2stars.get(repo_name):
                continue

            data = requests.get(repo_api).json()
            message = data.get("message", "")

            # Strategy: Skip task. API response is outdated, skip task.
            if "Not Found" in message:
                skip_repo.append(repo_name)
                continue
            # Strategy: Mission compete. IP address has been flagged.
            elif "limited" in message:
                break

            # Get statistics on contributed repository objects.
            repo2stars[repo_name] = data["stargazers_count"]
            updated_at = event["updated_at"]
            stick_url = event["html_url"]
            stick_alias = f"pull/{event['number']}"
            self.contributed_repo_objs[repo_name] = {
                "stars": repo2stars[repo_name],
                "updated_at": updated_at,
                "stick": f"[{stick_alias}]({stick_url})",
            }
            self.repo2hyperlink[repo_name] = stick_url


class MarkdownTemplater:
    def __init__(self):
        self.summary_context = []

    def update_summary(self, context: Union[str, list]):
        if isinstance(context, str):
            context = [context]
        for ctx in context:
            self.summary_context.append(ctx)

    def get_summary(self) -> str:
        return "\n".join(self.summary_context)

    def to_table(self, title: list, sequence: list) -> Optional[list]:
        try:
            if not title or not sequence or len(title) != len(sequence[0]):
                return []
        except (IndexError, TypeError):
            return []

        context = [
            f"|{'|'.join([f' **{t}** ' for t in title])}|",  # table title
            f"|{'|'.join([' :---- ', ] * len(sequence[0]))}|",  # table divider
        ]
        for s in sequence:
            context.append(f"|{'|'.join([str(i) for i in s])}|")  # table content

        self.update_summary(context)

        return context

    def to_title(self, text: str, level: int = 2) -> str:
        ctx = f"{'#' * level} {text}\n"
        self.update_summary(ctx)
        return ctx

    def to_unordered_list(self, sequence: Union[str, dict, List[str]]) -> Optional[List[str]]:
        context = []
        if isinstance(sequence, str):
            context.append(f"- {sequence}")
        elif isinstance(sequence, list):
            context = [f"- {s}" for s in sequence]
        elif isinstance(sequence, dict):
            context = [f"- **{str(k)}**: {str(v)}\n" for k, v in sequence.items()]
        self.update_summary(context)
        return context

    def to_ordered_list(self, sequence: Union[str, List[str]]) -> Optional[List[str]]:
        if isinstance(sequence, str):
            sequence = [sequence]

        context = [f"{i + 1}. {s}" for i, s in enumerate(sequence)]
        self.update_summary(context)
        return context

    def to_html_img(self, alt, src) -> str:
        ctx = f'<img width="30%" src="{src}" alt="{alt}"/>\n'
        self.update_summary(ctx)
        return ctx

    def to_content(self, text: str) -> Optional[str]:
        self.update_summary(text)
        return text

    def sayhi(self) -> str:
        ctx = f"## Hi there :heart:, I'm {GITHUB_ACTOR}\n"
        self.update_summary(ctx)
        return ctx


def cache_summary(summary):
    with open(PATH_SUMMARY_OUTPUT, "w", encoding="utf8") as file:
        file.write(summary)


def register_user_data():
    """Get the corresponding user statistics."""
    gus = GitUserStatus()
    gus.get_repos_info()
    gus.get_contributed_repos()

    return gus


def arrange_summary():
    gus = register_user_data()
    mt = MarkdownTemplater()

    # SayHi
    mt.sayhi()
    mt.to_html_img(alt=GITHUB_ACTOR, src=gus.avatar_url)

    # == Statistics ==
    mt.to_title("Statistics", level=2)
    mt.to_content(f"> Automated deployment @ {str(datetime.now()).split('.')[0]} Asia/Shanghai")

    # === User Statistics ===
    mt.to_title("User", level=3)
    # -- Arrange content --
    _user_statistics = {"Total_stars_earned": gus.total_stars_earned, "Projects": gus.public_repos}
    # -- Write content --
    mt.to_unordered_list(_user_statistics)

    # === Repositories Statistics ===
    mt.to_title("Sources", level=3)
    # -- Arrange content --
    repo_obj = sorted(
        gus.repo_objs.items(), key=lambda kv: (kv[-1]["pushed_at"], kv[0]), reverse=True
    )
    title = ["name"] + list(repo_obj[0][-1].keys())
    sequence = [[f"[{k}]({gus.repo2hyperlink[k]})"] + list(v.values()) for k, v in repo_obj]
    # -- Write content --
    mt.to_unordered_list({"count": len(sequence)})
    mt.to_table(title=title, sequence=sequence)

    # === Forked ===
    mt.to_title("Forks", level=3)
    # -- Arrange content --
    repo_obj = sorted(
        gus.forked_repo_objs.items(), key=lambda kv: (kv[-1]["pushed_at"], kv[0]), reverse=True
    )
    title = ["name"] + list(repo_obj[0][-1].keys())
    sequence = [[f"[{k}]({gus.repo2hyperlink[k]})"] + list(v.values()) for k, v in repo_obj]
    # -- Write content --
    mt.to_unordered_list({"count": len(sequence)})
    mt.to_table(title=title, sequence=sequence)

    # === Contributed ===
    mt.to_title("Contributed", level=3)
    # -- Arrange content --
    repo_obj = sorted(
        gus.contributed_repo_objs.items(), key=lambda kv: (kv[-1]["stars"], kv[0]), reverse=True
    )
    title = ["name"] + list(repo_obj[0][-1].keys())
    sequence = [[f"[{k}]({gus.repo2hyperlink[k]})"] + list(v.values()) for k, v in repo_obj]
    # -- Write content --
    mt.to_unordered_list({"count": len(sequence)})
    mt.to_table(title=title, sequence=sequence)

    # Output
    cache_summary(mt.get_summary())


if __name__ == "__main__":
    arrange_summary()
